### PR TITLE
Allow setting of compiler options

### PR DESCRIPTION
**Pull request by ijc**
To allow e.g. building a static binary with:

    make binaries GO_GCFLAGS="--ldflags '-extldflags \"-fno-PIC -static\"'"

Signed-off-by: Ian Campbell <ijc@docker.com>

### CODE REVIEW

**Comment by ijc** [comment]
I believe I have now signed the CLA. This comment should retrigger the check.

**Comment by ijc** [comment]
@mikebrow Happy to do so. Getting the quoting right for the echo to print exactly the right thing is pretty tricky. How would you feel about just adding a `static-binaries` target which just recurses to `binaries` with the right flags?

Self-nit: The example should really be using `GO_LDFLAGS` not `GO_CFLAGS` I think.

**Comment by mikebrow** [comment]
Dunno.. I sort of prefer GO_GCFLAGS over GO_CFLAGS... and containerd/containerd uses GO_GCFLAGS. Presume the goal here is a static of all 3, runc, containerd, and cri-containerd?

Yes, a static or static-binaries target would be nice. In that case may only need the LD flags to be used as and environment variable.  Just thinking out loud.

**Comment by ijc** [comment]
> Presume the goal here is a static of all 3, `runc`, `containerd`, and `cri-containerd`?

Correct. The context is [LinuxKit](https://github.com/linuxkit/linuxkit) and in particular the [projects/kubernetes](https://github.com/linuxkit/linuxkit/tree/master/projects/kubernetes) in there which I'd like be using `cri-containerd` (at least as an option).

FWIW LinuxKit's containerd build is using:
```
make binaries EXTRA_FLAGS="-buildmode pie" EXTRA_LDFLAGS="-extldflags \\\"-fno-PIC -static\\\""
```
Where `EXTRA_FLAGS` is passed to `go build` directly and `EXTRA_LDFLAGS` is appended to `GO_LDFLAGS`. It's quite the proliferation of options!

I'll go ahead and add a `static-binaries` target, but if you prefer I could do a slightly larger change modelled on containerd's use of all those flag names.

File: Makefile [hunk 1]
```diff
@@ -83,6 +85,9 @@ clean:
 
 binaries: cri-containerd
 
+static-binaries: GO_LDFLAGS=--ldflags '-extldflags "-fno-PIC -static"'
+static-binaries: cri-containerd
```
Comment (Random-Liu) [inline Makefile]: Add to `.PHONY`?

**Comment by Random-Liu** [comment]
> Correct. The context is LinuxKit and in particular the projects/kubernetes in there which I'd like be using cri-containerd (at least as an option).

@ijc I'm interested in this part. LInuxKit generated OS is using containerd to manage services, right? Will kubernetes inside talk to this containerd? Just curious. :)

File: Makefile [hunk 2]
```diff
@@ -83,6 +85,9 @@ clean:
 
 binaries: cri-containerd
 
+static-binaries: GO_LDFLAGS=--ldflags '-extldflags "-fno-PIC -static"'
```
Comment (mikebrow) [inline Makefile]: didn't realize where you were heading with this.. can we make these LD flags more specific and leave the environment variables to other uses?  say make your flags GO_STATIC_LDFLAGS?

Comment (Random-Liu) [inline Makefile]: We probably want to use `+=` here.

File: Makefile (hunk 1 above)
Comment (ijc) [inline Makefile]: Ack, will do.

File: Makefile (hunk 2 above)
Comment (ijc) [inline Makefile]: Shall I copy the `EXTRA_LDFLAGS` and `EXTRA_FLAGS` pattern from containerd perhaps?

**Comment by ijc** [comment]
> @ijc I'm interested in this part. LInuxKit generated OS is using containerd to manage services, right? Will kubernetes inside talk to this containerd? Just curious. :)

Yes, that's the plan, cri-containerd is running in a system container but with access to the system containerd control socket.

I have it working locally just need to clean up the patch and submit it to linuxkit. Currently I'm preempted with something else (completely unrelated) but I hope to get back to it in the next couple of days and post something. This PR isn't a blocker for my sending an initial patch (but it would be a nice to have of course!).

**Comment by Random-Liu** [comment]
> Yes, that's the plan, cri-containerd is running in a system container but with access to the system containerd control socket.

@ijc Thanks! Good to know.

**Comment by Random-Liu** [comment]
LGTM

**Pull request merged**